File: 144.binary-tree-preorder-traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        _preorderTraversal(root);
        return result;
    }
private:
    void _preorderTraversal(TreeNode* root) {
        if (!root) return;
        result.push_back(root->val);
        _preorderTraversal(root->left);
        _preorderTraversal(root->right);
    }
    vector<int> result;
};

File: 144.binary-tree-preorder-traversal/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
	var cache []int
	return preorderTraversal2(root, cache)
}

func preorderTraversal2(root *TreeNode, cache []int) []int {
	if root == nil {
		return cache
	}
	cache = append(cache, root.Val)
	cache = preorderTraversal2(root.Left, cache)
	return preorderTraversal2(root.Right, cache)
}
